iceoryx_examples/automotive_soa/include/owl/kom/event_publisher.hpp

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_EXAMPLES_AUTOMOTIVE_SOA_EVENT_PUBLISHER_HPP
+#define IOX_EXAMPLES_AUTOMOTIVE_SOA_EVENT_PUBLISHER_HPP
+
+#include "iceoryx_posh/popo/publisher.hpp"
+
+#include "owl/types.hpp"
+
+namespace owl
+{
+namespace kom
+{
+template <typename T>
+class EventPublisher
+{
+  public:
+    using SampleType = T;
+
+    EventPublisher(const ServiceIdentifier& service,
+                   const InstanceIdentifier& instance,
+                   const EventIdentifier& event) noexcept;
+
+    // Deleted because of SamplePointer implementation capturing 'this' in Loan()
+    EventPublisher(const EventPublisher&) = delete;
+    EventPublisher(EventPublisher&&) = delete;
+    EventPublisher& operator=(const EventPublisher&) = delete;
+    EventPublisher& operator=(EventPublisher&&) = delete;
+
+    static constexpr uint64_t HISTORY_CAPACITY{1U};
+    static constexpr bool NOT_OFFERED_ON_CREATE{false};
+
+    void Offer() noexcept;
+    void StopOffer() noexcept;
+
+    bool Send(const SampleType& userSample) noexcept;
+    //! [EventPublisher zero-copy send]
+    void Send(SamplePointer<SampleType> userSamplePtr) noexcept;
+    //! [EventPublisher zero-copy send]
+
+    //! [EventPublisher allocate]
+    SamplePointer<SampleType> Loan() noexcept;
```

Comment:
You have a livetime issue here. See this code:
```cxx
SamplePointer<SampleType>&& acquireSample() {
  EventPublisher<Topic> publisher;
  return std::move(publisher.Loan());
}

int main() {
  auto sample = acquireSample();
}
```

In the end you can construct a lot of examples where the `EventPublisher` goes out of scope before `SamplePointer` does and then one has a problem. I think this is another problem with the automotive API where one has to state at least a big red warning that the `EventPublisher` has to live as least as long as the `SamplePointer`.

But just so that you are aware how bad this problem is, take a look at this evil bug:
```cxx
class MyClass {
  SamplePointer<SampleType> m_sample; 
  EventPublisher<Topic> m_publisher;
}
```

This ordering of member variables is evil, when the `MyClass` destructor destroys the member it does it in inverse order, first the `m_publisher` and then the `m_sample` and one has again undefined behavior. I would bet in a code review no one would have seen this.
There are a lot of bugs out there where applications cannot terminate cleanly and cause leaks or segfaults because of this issue.

Reply:
This is an issue with the iceoryx API and would be solved if we solve it for `popo::Sample`:  https://github.com/eclipse-iceoryx/iceoryx/issues/1378